tests/cover/test_conjecture_engine.py

```diff
@@ -1243,6 +1243,27 @@ def x(data):
     assert x == hbytes([0, 0, 2])
 
 
+def test_shrinking_block_pairs(monkeypatch):
```

Comment:
It might be worth adding some more tests before you do anything else, as there are two corner cases in the code right now that this test is failing to catch and looking at them might be informative.

The first thing I think would be useful is that currently your solution is only ever running the test for the smallest possible example, so it would be useful to have a test like this one but where the smallest possible example doesn't work. For example you could replace the `m == n + 2` branch with a `m == n + 2 and n >= 3`.

The other thing that would be useful to try would be one where the block shape changes. So you could have something like um....

```python

@run_to_buffer
def x(data):
    m = data.draw_bits(8)
    if m < 5:
        data.mark_invalid()
    if m == 5:
        data.mark_interesting()
    n = data.draw_bits(8)
    if m == n + 2:
        data.mark_interesting()
assert x == hbytes([5])
```

What happens here is that the shrinker is merrily shrinking away trying to lower `m` as much as possible, but when it hits the magic number `m = 5`, suddenly the `n` disappears out from under it. This will currently fail in the same way as the above test, but I think when you've fixed the above test it will fail in a new and interesting way because of how you're calculating the pairs.